### PR TITLE
Metadata file format changes (function.json)

### PR DESCRIPTION
**Pull request by mathewc**
Putting bindings in a flattened array, and introducing a new "BindingDirection" enum (in/out/inout)

### CODE REVIEW

**Comment by davidebbo** [comment]
I didn't look at all the function.json files, so if one or two are off I won't catch it. But I looked at the code change and it looks straightforward. :shipit:

File: function.json [hunk 1]
```diff
@@ -1,20 +1,17 @@
...
+    {
+      "type": "eventHubTrigger",
+      "direction": "in",
+      "connectionString": "%AzureWebJobsEventHubReceiver%",
+      "path": "%AzureWebJobsEventHubPath%"
+    },
+    {
+      "type": "eventHub",
+      "name": "output",
+      "direction": "in",
```
Comment (mathewc) [inline function.json]: should be 'out'

**Comment by davidebbo** [comment]
Note that we need a matching PR for https://github.com/Azure/azure-webjobs-sdk-templates/tree/master/Templates/BlobTrigger

**Comment by mathewc** [comment]
Yep, I'll send the PR for template updates

File: sample/QueueTrigger-Python/function.json [hunk 2]
```diff
@@ -1,25 +1,24 @@
...
         "name": "input",
+        "direction": "in",
         "tableName": "samples",
         "partitionKey": "samples",
         "take": 5,
-        "filter": "status eq '0'"
-      }
-    ],
-    "output": [
+        "filter": "status eq '0"
```
Comment (mathewc) [inline sample/QueueTrigger-Python/function.json]: fix: mistakenly removed ending quote

**Comment by christopheranderson** [comment]
We have no "trigger" enum, so we have to regex match "Trigger" to find the trigger for a "function.json"? I don't think this will negatively impact the devExp, but it seems like we'd be doing some hacky things for ux/tooling. We should likely include flags for trigger in the "binding.json" to make things more resilient to things like a spelling mistake.

This got me thinking, so I'm proposing an alternative as a thought experiment. Could we kill direction all together and then the names are just "NewBlob", "GetBlob", "PutBlob"? This seems more universally descriptive. It would support us eventually doing "NewBlobPostEvent" and "NewBlobDeleteEvent". The originally proposed model could theoretically support names as well, but it seems like the "direction" property isn't actually representative of any behavioral changes, it's just a less clear abstraction over specific events. Why not make it very specific what the Event Type is in the config? Let the UX, which helps provide hints towards the "direction" worry about the direction.

tl;dr - Should we just use a more descriptive event type and do away with behavior in the "function.json"?

**Comment by davidebbo** [comment]
@christopheranderson Yes, I could go with something along the lines of what you propose.

**Comment by davidebbo** [comment]
But I didn't mean to redo this PR. Let's discuss and come up with a plan for later. :)

**Comment by christopheranderson** [comment]
Yeah, I don't feel strongly about blocking progress on this model, but I wanted to put the idea forward. I think we might be describing things the wrong way the more I've gone back and forth on the "behavior/direction" issue.

**Comment by mathewc** [comment]
I don't understand all the "NewBlobPostEvent" stuff proposed above - probably something best discussed in person :) However, I do think that the direction enum as I have it makes a lot of sense - it's concise and doesn't serve too many purposes. Something like "behavior" that has "in", "out", "inout", "trigger" attempts to do to many things. A trigger is a degenerate case of always "in" - we can even default the direction for triggers so it doesn't have to be specified.

**Comment by davidebbo** [comment]
Let's discuss outside of this pull request :)
